### DISCUSSION TITLE
How do I prevent Telescope and NeoTree from changing directories?

### DISCUSSION THREAD

**Discussion by pprotas** [LazyVim/LazyVim#952] [Q&A] [answered]
Hello, I have noticed this behavior started occurring some time ago in LazyVim and now I'm actually getting really annoyed by it :P Unfortunately, I can't find the root cause.

So I have a folder structure like this:
```
my_full_stack_project
|- rails_project
    |- example.rb
    |- random Rails or Ruby related files and configuration
|- react_project
    |- example.js
    |- random React, WebPack, Node, etc. files and configuration
```
Basically, it's a folder that contains a Rails project with `.rb` files and a React project with `.js(x)/.ts(x)` files. If my active buffer is a `.rb` file, I can use Telescope and NeoTree to go to any file, no matter in which project it is. But If I choose a file in the `react_project` as my active buffer, for example `example.js`, I can not find any `.rb` files. So if I type `example.rb` in my Telescope search box, nothing shows up. Also, when I open up NeoTree, it is only scoped to `react_project`, so the folder structure looks like this:
```
react_project
|- example.js
```

The only workaround I found is to use `<leader>E` to open up NeoTree in the root dir, and then start searching in Telescope. But it's very annoying!

My hunch is that there is some kind of configuration in `react_project` that is recognized by LazyVim or my Eslint LSP, and for some reason it's forcing the directory of my entire neovim to be `react_project`. The reason I think that is because I don't have this problem with `rails_project` at all

**Desired behavior**
1. Open `my_full_stack_project`
2. Go to `example.js`
3. Open up Telescope with `<leader><leader>`
4. Look for `example.rb`
5. **You can successfully find `example.rb`**
6. Open up NeoTree with `<leader>e`
7. **NeoTree opens in the root dir** 

**Actual behavior**
1. Open `my_full_stack_project`
2. Go to `example.js`
3. Open up Telescope with `<leader><leader>`
4. Look for `example.rb`
5. **`example.rb` does not show up in the list**
6. Open up NeoTree with `<leader>e`
7. **NeoTree opens in `react_project` and doesn't show you `rails_project`**

**Comment by jyuan0** [comment]
It sounds like it's due to this code: https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/util/init.lua#L91

Note: this code sets up an `<a-c>` mapping in Telescope to change between the root/cwd directories.

**Reply by pprotas** [reply]
That's some good investigative work! In fact, the function `get_root` above seems responsible:
```lua
-- returns the root directory based on:
-- * lsp workspace folders
-- * lsp root_dir
-- * root pattern of filename of the current buffer
-- * root pattern of cwd
```

So I'm thinking that either the `eslint-lsp` or `tsserver` LSP is making LazyVim switch the directory :p

Is there a way to turn this off? I almost never want my editor to be switching directories automatically, I just want it to stay in the dir that I opened it in.

IMO, this should be reversed. So I should have to use a keybinding (instead of it being default behavior) to open Telescope/NeoTree in the LSP workspace

**Reply by pprotas** [reply]
Here is the line for the config that is causing the issues in NeoTree: https://github.com/LazyVim/LazyVim/blob/47111bd2db140aac32422752e07db7a1c5b74659/lua/lazyvim/plugins/editor.lua#LL13C24-L13C24

As you can see, it's also using `get_root`

**Reply by jyuan0** [reply]
> Is there a way to turn this off? I almost never want my editor to be switching directories automatically, I just want it to stay in the dir that I opened it in.

There isn't a toggle or anything. Sounds like you want the default Telescope/Neo-tree behavior. If so, I think you should just remap the keys.

**Reply by jyuan0** [reply]
I think the `get_root` versions are also bound to `<leader>ff` / `<leader>fF`, so it's still available if you remap `<leader><leader>`.

**Reply by dpetka2001** [reply]
It's because those keymaps use `Util.telescope` function. This function uses the `Util.get_root` to decipher the root directory. If you want to not use the `Util.get_root` function in your keymaps, then you should just remap the combination of keys you want to use plain `Telescope find_files` instead I believe instead of `Util.telescope` like it already is in LazyVim.

**Reply by dpetka2001** [reply]
Or you could change the lsp-server `root_dir` setting to include a pattern (something like `.git` maybe?) so that it detects the top directory of your ruby and react projects (assuming that the very top directory is a git repo itself) or some other pattern. That would mean that both lsp-servers would be attached for both your `.rb` and `.js` though (I believe). I'm not sure you want this.

**Comment by folke** [comment, marked as answer]
If you don't want the root dir functionality, do the following:

```lua
require("lazyvim.util").get_root = vim.loop.cwd
```

**Reply by pprotas** [reply]
Perfect, this works just like expected now!

**Reply by robsonpeixoto** [reply]
@folke, thanks!

Just one more question. Where is the best place to add this config? I put it on `config/lazy.lua`:
```lua
❯ git diff lua/config/lazy.lua

Δ .config/nvim/lua/config/lazy.lua
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

───────────────────────────────────────────────────┐
• 9: vim.opt.rtp:prepend(vim.env.LAZY or lazypath) │
───────────────────────────────────────────────────┘
require("lazy").setup({
  spec = {
    -- add LazyVim and import its plugins
    { "LazyVim/LazyVim", import = "lazyvim.plugins" },
    {
      "LazyVim/LazyVim",
      import = "lazyvim.plugins",
      config = function()
        require("lazyvim.util").get_root = vim.loop.cwd
      end,
    },
    -- import any extras modules here
    -- { import = "lazyvim.plugins.extras.lang.typescript" },
    -- { import = "lazyvim.plugins.extras.lang.json" },

```

**Reply by folke** [reply]
Not in lazyvim config, since that will prevent lazyvim to properly setup.

lazyvim `init` should be fine

**Reply by dvic** [reply]
Thanks, was also looking for this. And how do we get rid of the obsolete mapping? (and change the description of the other one) I tried settings the `keys` field in `opts` of the `telescope` plugin but this does not work (and I'm not sure how to delete it).

**Reply by jyuan0** [reply]
You should be able to disable/change the `telescope` mappings in the `keys` field of the spec.

**Reply by dvic** [reply]
Hmm, I'm going to try again then, but just to be sure I try the thing that should work: how do you disable a mapping? use tuple `{"<leader>fF ,false}`?

**Reply by dpetka2001** [reply]
You're missing a quote at the end of your `lhs` expression. I hope it's just a typo and you don't actually have it like that in your spec. That's how I did it as well to disable some keybindings and move them to a different `which-key` namespace.

**Reply by jyuan0** [reply]
Something like this should work:
```
  {
    "nvim-telescope/telescope.nvim",
    keys = {
      { "<leader>ff", "<cmd>Telescope find_files<cr>", desc = "Find files" },
      { "<leader>fF", false},
    },
  },
```

**Reply by dvic** [reply]
Thanks that works! My mistake was to put the `keys` inside the `opts`, did not realize this was a property for lazy.nvim.

**Reply by carleryd** [reply]
`require("lazyvim.util").get_root` is deprecated and changing this had no effect, so I had to change `require("lazyvim.util").root.get` instead.

In my `~/.config/lazy.lua`
```
vim.api.nvim_create_autocmd("User", {
  pattern = "VeryLazy",
  callback = function()
    require("lazyvim.util").root.get = vim.loop.cwd
  end,
})
```
